scripts/keystone_data.sh

```diff
@@ -1,156 +1,395 @@
+
+    # invisible tenant - admin can't see this one
+    local invis_tenant
+    invis_tenant=$(get_or_create_project "invisible_to_admin" default)
+
+    # demo
+    local demo_tenant
+    demo_tenant=$(get_or_create_project "demo" default)
+    local demo_user
+    demo_user=$(get_or_create_user "demo" \
+        "$ADMIN_PASSWORD" "default" "demo@example.com")
+
+    get_or_add_user_project_role $member_role $demo_user $demo_tenant
+    get_or_add_user_project_role $admin_role $admin_user $demo_tenant
+    get_or_add_user_project_role $another_role $demo_user $demo_tenant
+    get_or_add_user_project_role $member_role $demo_user $invis_tenant
+
+    # alt_demo
+    local alt_demo_tenant
+    alt_demo_tenant=$(get_or_create_project "alt_demo" default)
+    local alt_demo_user
+    alt_demo_user=$(get_or_create_user "alt_demo" \
+        "$ADMIN_PASSWORD" "default" "alt_demo@example.com")
+
+    get_or_add_user_project_role $member_role $alt_demo_user $alt_demo_tenant
+    get_or_add_user_project_role $admin_role $admin_user $alt_demo_tenant
+    get_or_add_user_project_role $another_role $alt_demo_user $alt_demo_tenant
+
+    # groups
+    local admin_group
+    admin_group=$(get_or_create_group "admins" \
+        "default" "openstack admin group")
+    local non_admin_group
+    non_admin_group=$(get_or_create_group "nonadmins" \
+        "default" "non-admin group")
+
+    get_or_add_group_project_role $member_role $non_admin_group $demo_tenant
+    get_or_add_group_project_role $another_role $non_admin_group $demo_tenant
+    get_or_add_group_project_role $member_role $non_admin_group $alt_demo_tenant
+    get_or_add_group_project_role $another_role $non_admin_group $alt_demo_tenant
+    get_or_add_group_project_role $admin_role $admin_group $admin_tenant
+}
+
+#################################################
+# script starts here
+#################################################
+
+set -e
 
 ADMIN_PASSWORD=${ADMIN_PASSWORD:-secrete}
 SERVICE_PASSWORD=${SERVICE_PASSWORD:-$ADMIN_PASSWORD}
-export SERVICE_TOKEN=$SERVICE_TOKEN
-export SERVICE_ENDPOINT=$SERVICE_ENDPOINT
+SERVICE_ENDPOINT=$SERVICE_ENDPOINT
 SERVICE_TENANT_NAME=${SERVICE_TENANT_NAME:-service}
+SERVICE_PROJECT_NAME=${SERVICE_TENANT_NAME:-service}
+
+# bootstrap keystone (also creates endpoint and service in catalog)
+keystone-manage bootstrap \
+                --bootstrap-username admin \
+                --bootstrap-password "$ADMIN_PASSWORD" \
+                --bootstrap-project-name admin \
+                --bootstrap-role-name admin \
+                --bootstrap-service-name keystone \
+                --bootstrap-region-id "RegionOne" \
+                --bootstrap-admin-url "http://$SERVICE_HOST:\$(admin_port)s/v3" \
+                --bootstrap-public-url "http://$SERVICE_HOST:\$(public_port)s/v3" \
+                --bootstrap-internal-url "http://$SERVICE_HOST:\$(public_port)s/v3"
+
+set -x
+
+cat > $HOME/.userrc_early <<EOF
```

Comment:
can we simply export that directly instead of creating the file and sourcing it? is there really a lot of use for that file?